### PR TITLE
[Frontend][AST][IRGen] Improve availability support.

### PR DESCRIPTION
**Pull request by al45tair**
Use .def files to generate code for feature availability and runtime version to OS version mappings.

Also add a min-runtime-version option that can be used to avoid problems when building on Linux and Windows where because the runtime isn't part of the OS, availability doesn't solve the problem of trying to build the compiler against an older runtime.

Added some functions to IRGen to help with that problem; essentially we test *both* the deployment target availability *and* the new runtime version based availability.

rdar://121522431

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test macOS platform

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Review by eeckstein — APPROVED** [review]
Very nice, I like this!

Do I understand correctly that this PR contains of two separate features:
1. the new mechanism in the compiler for availability checking
2. passing the runtime version via build-script when building the compiler

(maybe worth splitting in two commits)

**Review by mikeash — APPROVED** [review]
Those .def files make everything so much nicer now.

**Comment by al45tair** [comment]
> Do I understand correctly that this PR contains of two separate features:
> 
> 1. the new mechanism in the compiler for availability checking
> 2. passing the runtime version via build-script when building the compiler

Yes. I suppose there are logically three pieces:

1. The new .def-file based availability stuff.
2. The `-min-runtime-version` compiler flag (plus the new functions in IRGen that are enabled by (1) that check it).
3. The use of (2) in the `SwiftCompilerSources` `CMakeLists.txt`.

I guess I could split it into three commits maybe to cover those?

**Comment by al45tair** [comment]
OK, I've broken it up into three separate commits.

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
I've also just realised that I somehow accidentally removed the code owners from the review list. I don't believe I did that myself, mind — it happened when I promoted this PR from a draft. Given that I'm tinkering with the availability code here, it's worth anyone interested taking a glance.

**Review by tshortli — COMMENTED** [review]
Nice, this seems like a big improvement.

File: include/swift/AST/RuntimeVersions.def [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+RUNTIME_VERSION(
+  (5, 9),
+  PLATFORM(macOS,   (14, 0, 0))
+  PLATFORM(iOS,     (17, 0, 0))
+  PLATFORM(watchOS, (10, 0, 0))
+)
+
+RUNTIME_VERSION(
+  (5, 10),
+  FUTURE
```
[the remark this answers is not in the document]
Comment (al45tair) [reply include/swift/AST/RuntimeVersions.def]: Updated this PR to match.

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Review by tshortli — APPROVED** [review]

**Pull request merged**